File: generate_index.js
const fs = require("fs");
const path = require("path");

const baseDir = path.join(__dirname, "historical-reports");
const outputFile = path.join(__dirname, "index.html");

/**
 * Retrieves the list of reports in the historical-reports directory.
 * Each report is expected to be inside a folder with an `index.html` file.
 * @param {string} dir - The directory containing reports.
 * @returns {Array} - List of report objects with name and URL.
 */
function getReportList(dir) {
    let reports = [];
    fs.readdirSync(dir, { withFileTypes: true }).forEach(entry => {
        if (entry.isDirectory()) {
            const reportPath = path.join(dir, entry.name, "index.html");
            if (fs.existsSync(reportPath)) {
                reports.push({
                    name: entry.name,
                    url: `historical-reports/${entry.name}/index.html`,
                });
            }
        }
    });

    // Sort reports (latest first)
    reports.sort((a, b) => parseReportDate(b.name) - parseReportDate(a.name));

    return reports;
}

/**
 * Extracts a timestamp from a report folder name (assumed format: "_Month_DD_YYYY_HH_MM").
 * @param {string} reportName - Report folder name.
 * @returns {number} - Timestamp for sorting or 0 if format is invalid.
 */
function parseReportDate(reportName) {
    const match = reportName.match(/_(\w+)_(\d{1,2})_(\d{4})_(\d{2})_(\d{2})$/);
    if (!match) return 0;
    const [_, month, day, year, hour, minute] = match;
    const months = { Jan: 0, Feb: 1, Mar: 2, Apr: 3, May: 4, Jun: 5, Jul: 6, Aug: 7, Sep: 8, Oct: 9, Nov: 10, Dec: 11 };
    return new Date(year, months[month], day, hour, minute).getTime();
}

const reports = getReportList(baseDir);

/**
 * Generates an HTML index page listing all reports.
 * @param {Array} reports - List of report objects.
 * @returns {string} - HTML content.
 */
function generateHTML(reports) {
    return `<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Historical Reports</title>
<style>
    body { font-family: Arial, sans-serif; margin: 20px; }
    h1 { color: #333; }
    table { width: 100%; border-collapse: collapse; }
    th, td { padding: 10px; border: 1px solid #ccc; text-align: left; }
    th { background-color: #f4f4f4; }
    a { text-decoration: none; color: blue; }
    .latest { font-weight: bold; color: green; }
</style>
</head>
<body>
<h1>Historical Reports</h1>
<table>
<thead>
<tr>
    <th>Report</th>
</tr>
</thead>
<tbody>
${reports.map(report => `
<tr>
    <td><a href="${report.url}" class="${report.name === reports[0].name ? 'latest' : ''}">
        ${report.name} ${report.name === reports[0].name ? '⬅️ Latest' : ''}
    </a></td>
</tr>`).join('')}
</tbody>
</table>
</body>
</html>`;
}

// Generate and save the report index
const htmlContent = generateHTML(reports);
fs.writeFileSync(outputFile, htmlContent, "utf8");

console.log(`✅ Generated: ${outputFile}`);

File: generate_index.dart
import 'dart:io';
import 'package:path/path.dart' as p;

void main() async {
  final reportsDir = Directory('historical-reports');
  final outputFile = File('index.html');

  if (!reportsDir.existsSync()) {
    print('Error: historical-reports folder not found.');
    return;
  }

  final buffer = StringBuffer()
    ..writeln('<!DOCTYPE html>')
    ..writeln('<html lang="en">')
    ..writeln('<head>')
    ..writeln('<meta charset="UTF-8">')
    ..writeln('<meta name="viewport" content="width=device-width, initial-scale=1.0">')
    ..writeln('<title>Historical Reports</title>')
    ..writeln('<style>')
    ..writeln('body { background: #f9fafb; font-family: "Segoe UI", Tahoma, Geneva, Verdana, sans-serif; display: flex; justify-content: center; align-items: start; padding: 2em; }')
    ..writeln('.container { background: white; box-shadow: 0 4px 12px rgba(0,0,0,0.05); border-radius: 12px; padding: 2em; max-width: 600px; width: 100%; }')
    ..writeln('h1 { color: #222; text-align: center; }')
    ..writeln('ul { list-style: none; padding: 0; margin-top: 1em; }')
    ..writeln('li { margin: 0.8em 0; display: flex; align-items: center; }')
    ..writeln('a { text-decoration: none; color: #1a73e8; font-size: 1.05em; flex-grow: 1; }')
    ..writeln('a:hover { text-decoration: underline; color: #0c47b7; }')
    ..writeln('.badge { background: #ff5252; color: white; font-size: 0.75em; padding: 2px 6px; border-radius: 8px; margin-left: 8px; }')
    ..writeln('.footer { text-align: center; margin-top: 2em; font-size: 0.9em; color: #777; }')
    ..writeln('</style>')
    ..writeln('</head>')
    ..writeln('<body>')
    ..writeln('<div class="container">')
    ..writeln('<h1>Historical Reports</h1>')
    ..writeln('<p style="color: #555; font-size: 0.95em;">If you don\'t see the latest report, wait a few minutes — we might still be processing something.</p>')
    ..writeln('<ul>');

  final subDirs = reportsDir
      .listSync()
      .whereType<Directory>()
      .where((dir) => dir.path.contains(RegExp(r'report_')))
      .toList()
    ..sort((a, b) => b.path.compareTo(a.path));  // Newest first!

  int count = 0;

  for (var i = 0; i < subDirs.length; i++) {
    final dir = subDirs[i];
    final dirName = p.basename(dir.path);
    final indexFile = File(p.join(dir.path, 'index.html'));

    if (indexFile.existsSync()) {
      buffer.write('<li><a href="historical-reports/$dirName/index.html">$dirName</a>');
      if (i == 0) {
        buffer.write('<span class="badge">Latest</span>');
      }
      buffer.writeln('</li>');
      count++;
    }
  }

  if (count == 0) {
    buffer.writeln('<li>No reports found.</li>');
  }

  buffer
    ..writeln('</ul>')
    ..writeln('<div class="footer">Generated on ${DateTime.now().toLocal()}</div>')
    ..writeln('</div>')
    ..writeln('</body>')
    ..writeln('</html>');

  await outputFile.writeAsString(buffer.toString());

  print('✅ index.html generated with $count reports.');
}
